File: Day1-9/1.py
def taxi_dist(x, y):
    return abs(x) + abs(y)


def part1(commands):
    x = y = 0
    curr_dir = (0, 1)
    for c in commands:
        turn, dist = c[0], c[1:]
        dist = int(dist)
        curr_dir = ((curr_dir[1], -curr_dir[0]) if turn == 'R'
                    else (-curr_dir[1], curr_dir[0]))
        x += curr_dir[0] * dist
        y += curr_dir[1] * dist
    return taxi_dist(x, y)


def part2(commands):
    visited = {}
    x = y = 0
    curr_dir = (0, 1)
    for c in commands:
        turn, dist = c[0], c[1:]
        dist = int(dist)
        curr_dir = ((curr_dir[1], -curr_dir[0]) if turn == 'R'
                    else (-curr_dir[1], curr_dir[0]))
        for _ in range(dist):
            x += curr_dir[0]
            y += curr_dir[1]
            if (x, y) in visited:
                return taxi_dist(x, y)
            visited[(x, y)] = True

if __name__ == '__main__':
    # Unit Tests
    assert part1(['R2', 'L3']) == 5
    assert part1(['R2', 'R2', 'R2']) == 2
    assert part1(['R5', 'L5', 'R5', 'R3']) == 12

    assert part2(['R8', 'R4', 'R4', 'R8']) == 4

    # Puzzles
    with open('1.txt', 'r') as f:
        commands = f.read().split(', ')

    print("Part 1: {}".format(part1(commands)))
    print("Part 2: {}".format(part2(commands)))

File: Day1-9/1.hs
import Data.List.Split
import Data.List

import qualified Data.Text as Text



data Direction = R | L deriving Show

turn compass dir = mod (compass + case dir of R -> 1
                                              L -> -1) 4

move prev (dir, n) = case last prev of
                    (compass, x, y) -> case turn compass dir of
                        0 -> [(0, x, y + i) | i <- [1..n]]
                        1 -> [(1, x + i, y) | i <- [1..n]]
                        2 -> [(2, x, y - i) | i <- [1..n]]
                        3 -> [(3, x - i, y) | i <- [1..n]]

part1 steps = 
    foldr (++) [] $ scanl move [(0, 0, 0)] steps

posOf (_, x, y) = (x, y)

part2 res =
    firstDup $ map posOf res

taxiDist x y = abs(x) + abs(y)

firstDup (x:xs) =
    case elem x xs of
        True -> x
        False -> firstDup xs

main = do
    input <- readFile "1.txt"
    let steps = [case head str of
                                'L' -> (L, read (tail str) :: Int)
                                'R' -> (R, read (tail str) :: Int)
                | str <- splitOn ", " input]
    let path = part1 steps
    let (_, x, y) = last path
    let p1res = taxiDist x y
    print $ "Part 1: " ++ show p1res
    let (dx, dy) = part2 path
    let p2res = taxiDist dx dy
    print $ "Part 2: " ++ show p2res
